### PR TITLE
Fix error 500 caused by ActionController::RoutingError (via fail-safe) when POST parameters are invalid

### LINKED ISSUE #29983: ActionController::RoutingError returns 500 (via fail-safe) rather than the expected 400 when POST parameters are invalid
I found the issue while investigating some 500 error logs in Unicorn.

This is the original stack-trace from unicorn:

```
Error during failsafe response: 822: unexpected token at '{record:{content:24.12.1.146}}'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/http/parameters.rb:88:in `rescue in parse_formatted_parameters'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/http/parameters.rb:82:in `parse_formatted_parameters'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/http/request.rb:354:in `block in POST'
  /var/www/shared/bundle/ruby/2.3.0/gems/rack-2.0.3/lib/rack/request.rb:57:in `fetch'
  /var/www/shared/bundle/ruby/2.3.0/gems/rack-2.0.3/lib/rack/request.rb:57:in `fetch_header'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/http/request.rb:353:in `POST'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/http/parameters.rb:45:in `parameters'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/http/mime_negotiation.rb:63:in `block in formats'
  /var/www/shared/bundle/ruby/2.3.0/gems/rack-2.0.3/lib/rack/request.rb:57:in `fetch'
  /var/www/shared/bundle/ruby/2.3.0/gems/rack-2.0.3/lib/rack/request.rb:57:in `fetch_header'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/http/mime_negotiation.rb:61:in `formats'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/middleware/public_exceptions.rb:22:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/middleware/show_exceptions.rb:49:in `render_exception'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/middleware/show_exceptions.rb:34:in `rescue in call'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/middleware/show_exceptions.rb:30:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/lograge-0.5.1/lib/lograge/rails_ext/rack/logger.rb:15:in `call_app'
  /var/www/shared/bundle/ruby/2.3.0/gems/railties-5.0.4/lib/rails/rack/logger.rb:24:in `block in call'
  /var/www/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.4/lib/active_support/tagged_logging.rb:69:in `block in tagged'
  /var/www/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.4/lib/active_support/tagged_logging.rb:26:in `tagged'
  /var/www/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.4/lib/active_support/tagged_logging.rb:69:in `tagged'
  /var/www/shared/bundle/ruby/2.3.0/gems/railties-5.0.4/lib/rails/rack/logger.rb:24:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/middleware/request_id.rb:24:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/rack-2.0.3/lib/rack/method_override.rb:22:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/rack-2.0.3/lib/rack/runtime.rb:22:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/activesupport-5.0.4/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/actionpack-5.0.4/lib/action_dispatch/middleware/executor.rb:12:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/rack-2.0.3/lib/rack/sendfile.rb:111:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/secure_headers-3.6.7/lib/secure_headers/middleware.rb:12:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/railties-5.0.4/lib/rails/engine.rb:522:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/newrelic_rpm-4.3.0.335/lib/new_relic/agent/instrumentation/middleware_tracing.rb:92:in `call'
  /var/www/shared/bundle/ruby/2.3.0/gems/unicorn-5.3.0/lib/unicorn/http_server.rb:606:in `process_client'
  /var/www/shared/bundle/ruby/2.3.0/gems/unicorn-5.3.0/lib/unicorn/http_server.rb:702:in `worker_loop'
  /var/www/shared/bundle/ruby/2.3.0/gems/unicorn-5.3.0/lib/unicorn/http_server.rb:549:in `spawn_missing_workers'
  /var/www/shared/bundle/ruby/2.3.0/gems/unicorn-5.3.0/lib/unicorn/http_server.rb:563:in `maintain_worker_count'
  /var/www/shared/bundle/ruby/2.3.0/gems/unicorn-5.3.0/lib/unicorn/http_server.rb:293:in `join'
  /var/www/shared/bundle/ruby/2.3.0/gems/unicorn-5.3.0/bin/unicorn:126:in `<top (required)>'
  /var/www/shared/bundle/bin/unicorn:17:in `load'
  /var/www/shared/bundle/bin/unicorn:17:in `<main>'
```

and this is the original error that triggered the fail-safe.

```
ActionController::RoutingError
No route matches [POST] "/domains/example-12334343333.com/records/72041.json"
```

I was also able to research a possible solution. I can submit a patch if you agree on the issue/solution.

1. when the exception occurs, `ActionDispatch::ShowExceptions` middleware is called
1. the middleware tries to render the exception via `render_exception(request, exception)` using the default exception app `ActionDispatch::PublicExceptions`
1. `ActionDispatch::PublicExceptions` tries to determine the content type for the response

    ```ruby
    content_type = request.formats.first
    ```

1. `request.formats` calls the `MimeNegotiation` module to extract the request formats

     ```ruby
        fetch_header("action_dispatch.request.formats") do |k|
          params_readable = begin
                              parameters[:format]
                            rescue ActionController::BadRequest
                              false
                            end
     ```

1. `parameters[:format]` crashes, because `parse_formatted_parameters` can't parse the POST data

    ```
      def parse_formatted_parameters(parsers)
        return yield if content_length.zero? || content_mime_type.nil?

        strategy = parsers.fetch(content_mime_type.symbol) { return yield }

        begin
          strategy.call(raw_post)
        rescue # JSON or Ruby code block errors
          my_logger = logger || ActiveSupport::Logger.new($stderr)
          my_logger.debug "Error occurred while parsing request parameters.\nContents:\n\n#{raw_post}"

          raise ParamsParser::ParseError
        end
      end
    ```

    I can confirm the crash is here because I can find the corresponding message in the logs

    ```
    Error occurred while parsing request parameters.
    Contents:
    {record:{content:24.12.1.146}}
    ```

1. As a result, a `ParamsParser::ParseError` is raised. But the `MimeNegotiation` module only rescues `ActionController::BadRequest` (and not `ParseError`). It looks in a normal request parameters paring errors may be re-packaged in an `ActionController::BadRequest`, but not in this case because the error is raised at a different level

1. As a result, the exception is bubbled up to the original `ShowException#render_exception` method, and rescued by the generic fail-safe mechanism

    ```ruby
    rescue Exception => failsafe_error
      $stderr.puts "Error during failsafe response: #{failsafe_error}\n  #{failsafe_error.backtrace * "\n  "}"
      FAILSAFE_RESPONSE
    ```

### Steps to reproduce

(Guidelines for creating a bug report are [available
here](http://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#creating-a-bug-report))

You can reproduce the error by sending a JSON POST request to a non-existing route with some mailformed payload.

```
curl -X POST \
  http://railsapp.host/missing.json \
  -H 'accept: application/json' \
  -H 'cache-control: no-cache' \
  -H 'content-type: application/json' \
  -d '{record:{content:24.12.1.146}}'
```

The result is

```
500 Internal Server Error
If you are the administrator of this website, then please read this web application's log file and/or the web server's log file to find out what went wrong.%
```

### Expected behavior

The request should properly handle the routing error and return 404.

### Actual behavior

The request crashes and the result is a 500 error.

### System configuration

**Rails version**:

5.0.4 but 5.1 is affected as well.

**Ruby version**:

2.3

### Proposed solution

The proposal is to change the following code

     ```ruby
        fetch_header("action_dispatch.request.formats") do |k|
          params_readable = begin
                              parameters[:format]
                            rescue ActionController::BadRequest
                              false
                            end
     ```

to relax the rescue in the mime negotiation to rescue any `StandardError` and set format to `false`.

Alternatively, the rescue statement should be changed to rescue from `ParserError`. I could not find a case where the parsing would raise an `ActionController::BadRequest` error.

There are a few cases where a post request can raise an `ActionController::BadRequest` as a result of a parsing:

```ruby
    def POST
      fetch_header("action_dispatch.request.request_parameters") do
        pr = parse_formatted_parameters(params_parsers) do |params|
          super || {}
        end
        self.request_parameters = Request::Utils.normalize_encode_params(pr)
      end
    rescue ParamsParser::ParseError # one of the parse strategies blew up
      self.request_parameters = Request::Utils.normalize_encode_params(super || {})
      raise
    rescue Rack::Utils::ParameterTypeError, Rack::Utils::InvalidParameterError => e
      raise ActionController::BadRequest.new("Invalid request parameters: #{e.message}")
    end
    alias :request_parameters :POST
```

but that's not the case of the `ParamsParser::ParseError` that is raised verbatim and in fact it's rescued in the ExceptionWrapper as it is:

```
      'ActionDispatch::ParamsParser::ParseError'      => :bad_request,
```

### PR DESCRIPTION
**Pull request by weppos**
### Summary

This PR fixes #29983 that I reported a few hours go.

In summary, when a request for an invalid route (that triggers a `ActionController::RoutingError`) contains invalid POST parameters (that can't be parsed), Rails crashes when trying to display the error and it returns an error 500 (because of the fail-safe mechanism) instead of a 404.

The ticket #29983 contains all the details, along with the backtraces of the bug I found in the app logs.

### Other Information

This PR includes a test to reproduce the error and avoid regressions. In #29983 I originally suggested to relax the rescue to StandardError, but instead I decided to add the specific exceptions in the rescue statement.

I will be happy to use a StandardError if you prefer.

### CODE REVIEW

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb [hunk 1]
```diff
@@ -60,7 +60,7 @@ def formats
         fetch_header("action_dispatch.request.formats") do |k|
           params_readable = begin
                               parameters[:format]
-                            rescue ActionController::BadRequest
+                            rescue ActionController::BadRequest, Parameters::ParseError
```
Comment (gmcgibbon) [inline actionpack/lib/action_dispatch/http/mime_negotiation.rb]: Maybe we should put it somewhere in here instead?

https://github.com/rails/rails/blob/98c1432583d4607af1467425183d8d448a692c5a/actionpack/lib/action_dispatch/http/parameters.rb#L54-L58

File: actionpack/test/dispatch/request_test.rb [hunk 2]
```diff
@@ -874,12 +874,28 @@ class RequestFormat < BaseRequestTest
...
+      "CONTENT_LENGTH" => body.length,
+      "CONTENT_TYPE" => "application/json",
+      "rack.input" => StringIO.new(body),
+      "action_dispatch.logger" => Logger.new(output = StringIO.new)
+    )
+    assert request.formats
+    assert request.format.html?
+
+    output.rewind && (err = output.read)
+    assert err =~ /Error occurred while parsing request parameters/
```
Comment (gmcgibbon) [inline actionpack/test/dispatch/request_test.rb]: You can do `assert_match` here instead for better failures.

**Review by gmcgibbon — COMMENTED** [review]
Still happens on `6.0.0.alpha`. Also worth noting this doesn't happen in development mode, only production. What environment setting adds mime negotiation?

File: actionpack/test/dispatch/request_test.rb (hunk 2 above)
Comment (weppos) [reply actionpack/test/dispatch/request_test.rb]: ✅  changed

File: actionpack/lib/action_dispatch/http/mime_negotiation.rb (hunk 1 above)
Comment (weppos) [reply actionpack/lib/action_dispatch/http/mime_negotiation.rb]: @gmcgibbon you knows it better than me. Feel free to adjust the PR.

**Comment by weppos** [comment]
@gmcgibbon I applied the suggested changes. There's one left, which may be beyond my knowledge of the implementation.

Any chance you can consider this PR for 6.x? It has been open for a while and we keep having a significant amount of logs generated.

**Comment by rafaelfranca** [comment]
Needs a rebase, but looks good to me. Can you rebase it?

**Comment by weppos** [comment]
@rafaelfranca thanks! I've rebased with master.

Actually, it looks like the bug was fixed in 6b3faf8e502dbd238fe4b4c409e96308638297a1, when the PR #34341 was merged. There was a change to use `RESCUABLE_MIME_FORMAT_ERRORS` that now includes this exception too.

I've merged the code anyways, and the PR is now basically just a test that protects against a regression.

Feel free to merge it, or close it if you don't think it's necessary.

**Pull request merged**